### PR TITLE
112  add trip stop sequence

### PR DESCRIPTION
**Pull request by jaygordon**
# Pull Request

Before submitting a pull request, please read [CONTRIBUTING.md](CONTRIBUTING.md).

For spec changes:

- [x] Pull-requests must address an existing issue
- [x] Update relevant documentation.

By contributing to this project, all contributors certify to the Developer Certificate of Origin in [CONTRIBUTING.md](CONTRIBUTING.md#contributor-agreement).

### CODE REVIEW

File: spec/fare_transactions.schema.json [hunk 1]
```diff
@@ -78,10 +78,18 @@
       "description": "Identifies the unique stop-path for a trip, may be distinct from GTFS shapes.shape_id"
     },
     {
-      "name": "stop_sequence",
+      "name": "trip_stop_sequence",
+      "type": "integer",
+      "description": "The temporal sequence of this transaction's enclosing stop visit, if any, within its performed trip, starting at 1 and containing no gaps. This may differ from scheduled_stop_sequence.",
```
Comment (botanize) [inline spec/fare_transactions.schema.json:83]: I find the description a bit unclear. Maybe something more akin to the GTFS stop_sequence language: 
```
"The actual order of stops visited within a performed trip. The values must start at 1 and must be consecutive along the trip.<br/> Example: A bus departs the first stop and detours around the second and third scheduled stops, visiting one unscheduled stop and resuming regular service at the 4th scheduled stop. The `scheduled_stop_sequence` is [1, null, 4], and the `trip_stop_sequence` is [1, 2, 3]."

File: spec/vehicle_locations.schema.json [hunk 2]
```diff
@@ -38,12 +38,21 @@
       "description": "Identifies the trip performed."
     },
     {
-      "name": "stop_sequence",
+      "name": "trip_stop_sequence",
+      "type": "integer",
+      "description": "The temporal sequence of the stop visit that the vehicle is serving or approaching within its performed trip, starting at 1 and containing no gaps. This may differ from scheduled_stop_sequence.",
+      "constraints": {
+        "required": true,
```
Comment (botanize) [inline spec/vehicle_locations.schema.json:45]: I see we're transferring the required constraint to `trip_stop_sequence`. However, if `trip_id_performed` isn't required I'm not sure how to make sense of `trip_stop_sequence` being required. Let's drop the required constraint on this field.

**Review by botanize — CHANGES_REQUESTED** [review]
- [ ] add `title` parameter to `trip_stop_sequence` field, in all tables, see comments for format.
- [ ] clarify `trip_stop_sequence` description property in all tables
- [ ] remove required constraint from `trip_stop_sequence` in `vehicle_locations` table

File: spec/fare_transactions.schema.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaygordon) [reply spec/fare_transactions.schema.json:84]: Added

File: spec/fare_transactions.schema.json [hunk 3]
```diff
@@ -78,10 +78,19 @@
       "description": "Identifies the unique stop-path for a trip, may be distinct from GTFS shapes.shape_id"
     },
     {
-      "name": "stop_sequence",
+      "name": "trip_stop_sequence",
+      "type": "integer",
+      "title": "Positive integer referencing stop_visits.trip_stop_sequence"
+      "description": "The actual order of stops visited within a performed trip. The values must start at 1 and must be consecutive along the trip.<br/> Example: A bus departs the first stop and detours around the second and third scheduled stops, visiting one unscheduled stop and resuming regular service at the 4th scheduled stop. The `scheduled_stop_sequence` is [1, null, 4], and the `trip_stop_sequence` is [1, 2, 3]."
```
Comment (botanize) [inline spec/fare_transactions.schema.json:84]: Missing a comma at the end of this line to separate `description` from `constraints`

File: spec/passenger_events.schema.json [hunk 4]
```diff
@@ -41,12 +41,21 @@
       }
     },
     {
-      "name": "stop_sequence",
+      "name": "trip_stop_sequence",
       "type": "integer",
-      "title": "positive integer referencing GTFS stop_times.stop_sequence",
-      "description": "Ordered stop the vehicle is serving on a particular trip. References GTFS",
+      "title": "Positive integer referencing stop_visits.trip_stop_sequence"
+      "description": "The actual order of stops visited within a performed trip. The values must start at 1 and must be consecutive along the trip.<br/> Example: A bus departs the first stop and detours around the second and third scheduled stops, visiting one unscheduled stop and resuming regular service at the 4th scheduled stop. The `scheduled_stop_sequence` is [1, null, 4], and the `trip_stop_sequence` is [1, 2, 3]."
```
Comment (botanize) [inline spec/passenger_events.schema.json:47]: Missing a comma at the end of this line to separate `description` from `constraints`

File: spec/stop_visits.schema.json [hunk 5]
```diff
@@ -29,12 +28,20 @@
       }
     },
     {
-      "name": "stop_sequence",
+      "name": "trip_stop_sequence",
       "type": "integer",
-      "title": "positive integer referencing GTFS stop_times.stop_sequence",
-      "description": "Ordered stop the vehicle is serving on a particular trip. References GTFS",
+      "description": "The actual order of stops visited within a performed trip. The values must start at 1 and must be consecutive along the trip.<br/> Example: A bus departs the first stop and detours around the second and third scheduled stops, visiting one unscheduled stop and resuming regular service at the 4th scheduled stop. The `scheduled_stop_sequence` is [1, null, 4], and the `trip_stop_sequence` is [1, 2, 3]."
```
Comment (botanize) [inline spec/stop_visits.schema.json:33]: Missing a comma at the end of this line to separate `description` from `constraints`

File: spec/vehicle_locations.schema.json [hunk 6]
```diff
@@ -38,12 +38,21 @@
       "description": "Identifies the trip performed."
     },
     {
-      "name": "stop_sequence",
+      "name": "trip_stop_sequence",
+      "type": "integer",
+      "title": "Positive integer referencing stop_visits.trip_stop_sequence"
+      "description": "The actual order of stops visited within a performed trip. The values must start at 1 and must be consecutive along the trip.<br/> Example: A bus departs the first stop and detours around the second and third scheduled stops, visiting one unscheduled stop and resuming regular service at the 4th scheduled stop. The `scheduled_stop_sequence` is [1, null, 4], and the `trip_stop_sequence` is [1, 2, 3]."
```
Comment (botanize) [inline spec/vehicle_locations.schema.json:44]: Missing a comma at the end of this line to separate `description` from `constraints`

**Review by botanize — CHANGES_REQUESTED** [review]
Thank you Jay! This looks good, but the JSON doesn't validate because there are some commas missing. I've noted the locations in the comments.

- [ ] add missing a comma at the end of the description property lines for the `trip_stop_sequence` field.

**Comment by e-lo** [comment]
@jaygordon can you also update the `architecture.md` mermaid diagram where applicable to reflect these term changes?

**Comment by jlstpaul** [comment]
@jaygordon If you have addressed @botanize comments, I think this can move forward.

**Comment by e-lo** [comment]
I think something like this would be fine:

```
stop_times.txt --> |"
        stop_times.txt: stop_id,stop_sequence
        stop_visits: stop_id,scheduled_stop_sequence
    "| stop_visits

```

```mermaid

graph LR;
stop_times.txt --> |"
        stop_times.txt: stop_id,stop_sequence
        stop_visits: stop_id,scheduled_stop_sequence
    "| stop_visits

```

Note: in [mermaid.live](https://mermaid.live/) you can use markdown which is prettier...but that doesn't seem to consistently render in github :-/

**Review by SorenSpicknall — APPROVED** [review]

**Pull request merged**